### DISCUSSION TITLE
MallocMessageBuilder for list types

### DISCUSSION THREAD

**Discussion by lewisxy** [capnproto/capnproto#1863] [Q&A]
What if I want to store a list of structs on a RPC server, or build list of structs in preparation for RPC? `capap::MallocMessageBuilder` does not seem to support list type at top level.
```
interface MyAPI {
  struct MyStruct {
    x @0 :UInt64;
    y @1 :UInt64;
  }

  useList @0 (input :List(MyStruct)) -> (result :List(MyStruct));
}
```
```c++
auto buf = capnp::MallocMessageBuilder();
auto listBuilder = buf.initRoot<capnp::List<MyAPI::MyStruct>>(); // won't compile, see below error
```
```
/opt/homebrew/Cellar/capnp/1.0.1/include/capnp/any.h:821:44: error: too few arguments to function call, expected 2, have 1
  return _::PointerHelpers<T>::init(builder);
         ~~~~~~~~~~~~~~~~~~~~~~~~~~        ^
/opt/homebrew/Cellar/capnp/1.0.1/include/capnp/message.h:461:28: note: in instantiation of function template specialization 'capnp::AnyPointer::Builder::initAs<capnp::List<MyAPI::MyStruct, capnp::Kind::STRUCT>>' requested here
  return getRootInternal().initAs<RootType>();
                           ^
client.cpp:31:28: note: in instantiation of function template specialization 'capnp::MessageBuilder::initRoot<capnp::List<MyAPI::MyStruct, capnp::Kind::STRUCT>>' requested here
    auto listBuilder = buf.initRoot<capnp::List<MyAPI::MyStruct>>();
                           ^
/opt/homebrew/Cellar/capnp/1.0.1/include/capnp/pointer-helpers.h:91:43: note: 'init' declared here
  static inline typename List<T>::Builder init(PointerBuilder builder, uint size) {
                                          ^
```

**Comment by kentonv** [comment]
If you are just using the MessageBuilder for storage and not to serialize a message, then you can use orphans and leave the message's root empty.

```
capnp::MallocMessageBuilder message;
capnp::Orphan<capnp::List<MyAPI::MyStruct>> orphan =
    message.getOrphanage.newOrphan<capnp::List<MyAPI::MyStruct>>(size);
capnp::List<MyAPI::MyStruct>::Builder list = orphan.get();
```

Note that `orphan` owns the underlying list so much be kept alive while you have any readers or builders pointing into it. (When the orphan is destroyed, the structure is zero'd out in the message.)

**Reply by lewisxy** [reply]
While `orphan` owns the underlying object, I am assuming `message` also needs to be alive as it's the backing buffer of the `orphan` in the example you provided? When an `orphan` goes out of scope, what happen to the memory (in `MallocMessageBuilder`) it once used? Can they be reused for other purpose?

**Reply by kentonv** [reply]
> While orphan owns the underlying object, I am assuming message also needs to be alive as it's the backing buffer of the orphan in the example you provided?

Yes.

> When an orphan goes out of scope, what happen to the memory (in MallocMessageBuilder) it once used? Can they be reused for other purpose?

Unfortunately, the memory is merely zeroed out and never reused. The memory allocator for MessageBuilders is a simple bump allocator, which suits the usual use case of building a message that will be immediately written out. It's not sophisticated enough to figure out when a "hole" in the message left by a removed object could be reused for a new object. [This is discussed a bit in the docs](https://capnproto.org/cxx.html#tips-and-best-practices) (see fifth bullet point).
